processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -918,19 +970,59 @@ private static int compareDimsInRowsWithAggs(
         final List<DimensionSpec> dimensions,
         final long maxDictionarySize,
         final DefaultLimitSpec limitSpec,
-        final List<ValueType> valueTypes
+        final List<ValueType> valueTypes,
+        @Nullable final List<String> dictionary
     )
     {
       this.includeTimestamp = includeTimestamp;
       this.sortByDimsFirst = sortByDimsFirst;
       this.dimensions = dimensions;
       this.dimCount = dimensions.size();
-      this.maxDictionarySize = maxDictionarySize;
       this.valueTypes = valueTypes;
       this.limitSpec = limitSpec;
-      this.serdeHelpers = makeSerdeHelpers();
+      this.enableRuntimeDictionaryGeneration = dictionary == null;
+      this.dictionary = enableRuntimeDictionaryGeneration ? new ArrayList<>(DICTIONARY_INITIAL_CAPACITY) : dictionary;
+      this.reverseDictionary = enableRuntimeDictionaryGeneration ?
+                               new Object2IntOpenHashMap<>(DICTIONARY_INITIAL_CAPACITY) :
+                               new Object2IntOpenHashMap<>(dictionary.size());
+      this.reverseDictionary.defaultReturnValue(UNKNOWN_DICTIONARY_KEY);
+      this.maxDictionarySize = maxDictionarySize;
+      this.serdeHelpers = makeSerdeHelpers(limitSpec != null, enableRuntimeDictionaryGeneration);
       this.keySize = (includeTimestamp ? Longs.BYTES : 0) + getTotalKeySize();
       this.keyBuffer = ByteBuffer.allocate(keySize);
+
+      if (!enableRuntimeDictionaryGeneration) {
+        final long initialDictionarySize = dictionary.stream()
+                                                     .mapToLong(RowBasedGrouperHelper::estimateStringKeySize)
+                                                     .sum();
+        Preconditions.checkState(
+            maxDictionarySize >= initialDictionarySize,
+            "Dictionary size[%s] exceeds threshold[%s]",
+            initialDictionarySize,
+            maxDictionarySize
+        );
+
+        for (int i = 0; i < dictionary.size(); i++) {
+          reverseDictionary.put(dictionary.get(i), i);
+        }
+
+        initializeSortableIds();
+      }
+    }
+
+    private void initializeSortableIds()
+    {
+      final int dictionarySize = dictionary.size();
+      final Pair<String, Integer>[] dictAndIds = new Pair[dictionarySize];
```

Comment:
You can avoid creating this array, by creating `sortableIds` right away and sorting it. Also, maybe "sortedIds"?

Reply:
This is not so simple. sortableIds is used to get the order of dictionary ids without comparing strings. Please check bufferComparator in AbstractStringRowBasedKeySerdeHelper.

I'm not sure where its name come from. @gianm do you have any idea?

Reply:
I don't say you should remove `sortableIds`, I say that you should remove `dictAndIds`. This array is created and thrown away just inside this method. It seems to me that you could rewrite this method without creating this array.

Reply:
I'm also talking about dictAndIds. dictAndIds is needed to make sortableIds because the order of keys in dictionary should not be changed and sortableIds should be accessed using the original index of keys in dictionary. The idea behind is

- When a new key appears, it is appended to the dictionary. Its id is the index in the dictionary.
- The id of the key is stored in ByteBuffer in BufferHashGrouper. The order of keys in dictionary should not be changed to get the proper keys by using ids in ByteBuffer.
- When BufferHashGrouper sorts the key ids in ByteBuffer, they should be sorted by comparing the keys, not their ids. A naive implementation will be getting keys from dictionary using the ids in ByteBuffer and comparing them. But, here, sortableIds is used to compare dictionary keys without string compare. We can directly get the sort order from sortableIds using the ids of ByteBuffer.

Reply:
I renamed ```sortableIds``` to ```rankOfDictionaryIds```.

Reply:
l don't understand why the body of this method couldn't look like this:
```java
      final int dictionarySize = dictionary.size();
      rankOfDictionaryIds = IntStream.range(0, dictionarySize).toArray();
      IntArrays.quickSort(rankOfDictionaryIds, (i1, i2) -> dictionary.get(i1).compareTo(dictionary.get(i2)));
```
Plus you will need to update `fastutil` library to 8.1.0

Reply:
This change is not possible because ```rankOfDictionaryIds``` is no more accessible using the original dictionary ids (the ids before sorted). You can see GroupByQueryRunnerTest is failed.

Reply:
Ok, then after quickSort() a loop that inverses indices and values could follow, it is possible to do it in-place because it's the values is a permutation of the indices.

Reply:
I don't understand this comment. What do you mean by "a loop that inverses indices and values"? 
The ```rankOfDictionaryIds``` represents the ranks of strings in dictionary. For example, given a dictionary of ["b", "c", "a"], ```rankOfDictionaryIds``` should be [1, 2, 0].

Reply:
"Inverse" turns [2, 0, 1] into [1, 2, 0].
It's like
```java
for (int i = 0; i != 0;) {
  int nextI = a[i];
  a[i] = i;
  i = nextI;
}
```

Reply:
Ok, it's not that easy, because of stable points and sub-loops.
```java
void inverseLoop(int[] a, int startIndex, BitSet visited) {
  int i = startIndex;
  do {
    visited.set(i);
    int nextI = a[i];
    a[i] = i;
    i = nextI;
  } while (i != startIndex);
}

void inverse(int[] a) {
  BitSet visited = new BitSet(a.length);
  for (int i = 0; i < a.length; i++) {
    if (!visited.get(i)) {
      inverseLoop(a, i, visited);
    }
  }
}
```

Reply:
I don't understand what is the benefit of this change. The bottleneck here was sorting with a TreeMap which is mitigated with an array sort now.
Your suggestion looks to need less memory, but is complicated and I'm not sure this will be faster than my implementation.

Reply:
I suggested it in the first place because of memory. Remember that you allocate not just array, but also a lot of pair objects, and a lot of boxed `Integer` objects.

With `inverse()` put somewhere like "IntArrayUtils" and tested separately, this method (BTW it should be renamed, because there are no more "sortableIds") will condense to 4 lines instead of 11 now.

Reply:
Ok. I fixed a bug in the above code and changed.